### PR TITLE
Add code modifier compiler plugin implementation

### PR DESCRIPTION
**Pull request by pramodya1994**
- Fixes https://github.com/ballerina-platform/ballerina-lang/issues/35232

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeModifierManager.java [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+        }
+
+        private Package modifyModule(ModuleId moduleId, Package pkg, CodeModifyTaskResult codeModifyTaskResult) {
+            Module module = pkg.module(moduleId);
+            Module.Modifier modifier = module.modify();
+
+            for (DocumentId documentId : pkg.module(moduleId).documentIds()) {
+                for (ModifiedSourceFile sourceFile : codeModifyTaskResult.sourceFiles(documentId)) {
+                    Document document = module.document(documentId);
+                    updateModifiedDocument(document.name(), sourceFile.textDocument(), modifier, documentId);
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeModifierManager.java:458]: We can use `document.modify().withContent().apply()` here

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Package.java [hunk 2]
```diff
@@ -256,6 +256,48 @@ public CodeGeneratorResult runCodeGeneratorPlugins() {
...
+     *
+     *   // Compile the package with generated files.
+     *   PackageCompilation compilation = packageWithGenFiles.getCompilation();
+     *   </pre>
+     * <p>
+     * This method does not run other tasks such as {@code CodeAnalyzer}s in engaged compiler plugins.
+     *
+     * @return a {@code CodeModifierResult} instance
+     */
+    public CodeModifierResult runCodeModifierPlugins() {
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/Package.java:278]: We should try to run code modifiers and code generators in one go. Running these separately will compile the files multiple times which will increase the build time.

File: project-api/project-api-test/src/test/resources/compiler_plugin_tests/package_comp_plugin_code_modify_add_function/CompilerPlugin.toml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+[plugin]
+class = "io.samjs.plugins.init.codemodify.CodeModifyFunctionPlugin"
+
+[[dependency]]
+path = "../../../../../build/compiler-plugin-jars/init-function-code-modify-compiler-plugin-1.0.0.jar"
+
+[[dependency]]
+path = "../../../../../build/compiler-plugin-jars/diagnostic-utils-lib-1.0.0.jar"
```
Comment (azinneera) [inline project-api/project-api-test/src/test/resources/compiler_plugin_tests/package_comp_plugin_code_modify_add_function/CompilerPlugin.toml:8]: missing new line

File: project-api/project-api-test/src/test/resources/compiler_plugin_tests/package_plugin_code_modify_user_1/main.bal [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import samjs/package_comp_plugin_code_modify_add_function as _;
+
+public function main() {
+}
+
+function foo() {
+}
+
+function bar() {
+}
```
Comment (azinneera) [inline project-api/project-api-test/src/test/resources/compiler_plugin_tests/package_plugin_code_modify_user_1/main.bal:10]: missing new line

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeModifierManager.java [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        }
+
+        @Override
+        public PackageCompilation compilation() {
+            return compilation;
+        }
+
+        @Override
+        public void modifySourceFile(TextDocument textDocument, DocumentId documentId) {
+            if (this.currentPackage.project().kind().equals(ProjectKind.SINGLE_FILE_PROJECT)) {
```
[the remark this answers is not in the document]
Comment (pramodya1994) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeModifierManager.java:274]: Removed breaking logic for single bal files and added a test case.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/CompilerPluginContextIml.java [hunk 6]
```diff
@@ -51,6 +53,10 @@ public void addCodeGenerator(CodeGenerator codeGenerator) {
         codeGenerators.add(new CodeGeneratorManager.CodeGeneratorInfo(codeGenerator, compilerPluginInfo));
     }
 
+    public void addCodeModifiers(CodeModifier codeModifier) {
```
[the remark this answers is not in the document]
Comment (pramodya1994) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/CompilerPluginContextIml.java]: Added and renamed

File: project-api/project-api-test/src/test/java/io/ballerina/projects/test/plugins/CompilerPluginTests.java [hunk 7]
```diff
@@ -271,6 +276,48 @@ public void testCompilerPluginCodegenBasic() {
...
+        Assert.assertEquals(currentPackage.packageDependencies().size(), 1,
+                "Unexpected number of dependencies");
+
+        // Running the code generation
+        CodeModifierResult codeModifierResult = currentPackage.runCodeModifierPlugins();
+
+        // Compiling the new package
+        Project project = currentPackage.project();
+        Package newPackage = codeModifierResult.updatedPackage().orElse(null);
+        Assert.assertNotNull(newPackage, "Cannot be null, because there exist code generators");
```
[the remark this answers is not in the document]
Comment (pramodya1994) [reply project-api/project-api-test/src/test/java/io/ballerina/projects/test/plugins/CompilerPluginTests.java:298]: Added

File: project-api/project-api-test/src/test/resources/compiler_plugin_tests/package_comp_plugin_code_modify_add_function/CompilerPlugin.toml (hunk 3 above)
Comment (pramodya1994) [reply project-api/project-api-test/src/test/resources/compiler_plugin_tests/package_comp_plugin_code_modify_add_function/CompilerPlugin.toml:8]: Added

File: project-api/project-api-test/src/test/resources/compiler_plugin_tests/package_plugin_code_modify_user_1/main.bal (hunk 4 above)
Comment (pramodya1994) [reply project-api/project-api-test/src/test/resources/compiler_plugin_tests/package_plugin_code_modify_user_1/main.bal:10]: Added

File: project-api/test-artifacts/init-function-code-modify-compiler-plugin/build.gradle [hunk 8]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (pramodya1994) [reply project-api/test-artifacts/init-function-code-modify-compiler-plugin/build.gradle:2]: Added

**Review by sameerajayasoma — COMMENTED** [review]
I reviewed only the Public API of this PR. LGTM

**Comment by pramodya1994** [comment]
Merged to master via https://github.com/ballerina-platform/ballerina-lang/pull/35532

**Pull request closed without merging**